src/libs/elektra/key.c

```diff
@@ -245,217 +245,190 @@ Key * keyVNew (const char * name, va_list va)
- *     was passed for dest or a dynamic property could not
- *     be written. The content will be unmodified then.
- * @retval 0 when dest was cleaned
- * @retval 1 when source was successfully copied
- * @see keyDup() to get a duplication of a key
+ *
+ * @return @p dest or NULL, in case of error. Possible errors are:
+ *   - memory allocation problems
+ *   - a part of @p dest that should be modified (e.g. name, value) was marked read-only,
+ *     e.g. the name of @p dest will be read-only if @p dest is part of a KeySet
+ *   - @p dest is NULL
  */
-int keyCopy (Key * dest, const Key * source)
+Key * keyCopy (Key * dest, const Key * source, elektraCopyFlags flags)
 {
-	if (!dest) return -1;
+	if (dest == NULL) return NULL;
 
-	if (test_bit (dest->flags, KEY_FLAG_RO_NAME) || test_bit (dest->flags, KEY_FLAG_RO_VALUE) ||
-	    test_bit (dest->flags, KEY_FLAG_RO_META))
-	{
-		return -1;
-	}
+	if (test_bit (dest->flags, KEY_FLAG_RO_NAME) && test_bit (flags, KEY_CP_NAME)) return NULL;
+	if (test_bit (dest->flags, KEY_FLAG_RO_VALUE) && test_bit (flags, KEY_CP_VALUE)) return NULL;
+	if (test_bit (dest->flags, KEY_FLAG_RO_META) && test_bit (flags, KEY_CP_META)) return NULL;
+
+	if (test_bit (flags, KEY_CP_STRING) && test_bit (flags, KEY_CP_VALUE)) return NULL;
 
-	if (!source)
+	if (source == NULL)
 	{
-		keyClear (dest);
-		return 0;
+		if (test_bit (flags, KEY_CP_NAME))
+		{
+			keySetName (dest, "/");
+		}
+		if (test_bit (flags, KEY_CP_VALUE))
+		{
+			keySetRaw (dest, NULL, 0);
+		}
+		if (test_bit (flags, KEY_CP_META))
+		{
+			ksClear (dest->meta);
+		}
+		return dest;
 	}
 
-	// remember dynamic memory to be removed
-	char * destKey = dest->key;
-	char * destUKey = dest->ukey;
-	void * destData = dest->data.c;
-	KeySet * destMeta = dest->meta;
+	if (test_bit (flags, KEY_CP_STRING) && keyIsBinary (source)) return NULL;
 
-	// duplicate dynamic properties
-	if (source->key)
-	{
-		dest->key = elektraStrNDup (source->key, source->keySize);
-		if (!dest->key) goto memerror;
-	}
-	else
-	{
-		dest->key = 0;
-	}
+	if (source == dest) return dest;
 
-	if (source->ukey)
-	{
-		dest->ukey = elektraStrNDup (source->ukey, source->keyUSize);
-		if (!dest->ukey) goto memerror;
-	}
-	else
-	{
-		dest->ukey = 0;
-	}
+	// remember original data of dest
+	Key orig = *dest;
 
+	// TODO: check MMAP flags
 
-	if (source->data.v)
+	// duplicate dynamic properties
+	if (test_bit (flags, KEY_CP_NAME))
 	{
-		dest->data.v = elektraStrNDup (source->data.v, source->dataSize);
-		if (!dest->data.v) goto memerror;
+		if (source->key)
+		{
+			dest->key = elektraStrNDup (source->key, source->keySize);
+			if (!dest->key) goto memerror;
+			dest->keySize = source->keySize;
+
+			ELEKTRA_ASSERT (source->ukey != NULL, "key != NULL but ukey == NULL");
+			dest->ukey = elektraStrNDup (source->ukey, source->keyUSize);
+			if (!dest->ukey) goto memerror;
+			dest->keyUSize = source->keyUSize;
+		}
+		else
+		{
+			dest->key = elektraStrDup ("/");
```

Comment:
Why not use keySetName (dest, "/");? Or why not use the same code above?

Reply:
Yes, it would actually make sense to use the same code above. I didn't call `keySetName()` since we know that `"/"` is a valid keyname, so we can avoid the validation. On the other hand, always calling `keySetName()` (even for the copy case) could make sense in future, if we add optimizations for certain values to `keySetName()`.

But I'll leave it as is right now, since this code will change anyways when we remove the escaped name.

---

PS. For the next time: it would be helpful, if you could clarify what you're referring to with "above" e.g. by [adding a permalink](https://docs.github.com/en/github/managing-your-work-on-github/creating-a-permanent-link-to-a-code-snippet) like this:

https://github.com/ElektraInitiative/libelektra/blob/1ca536139a956b528de861403e6ded88b302b686/src/libs/elektra/key.c#L309